src/EditorFeatures/Core/Implementation/CodeDefinitionWindow/DefinitionContextTracker.cs

```diff
@@ -0,0 +1,240 @@
+            ICodeDefinitionWindowService codeDefinitionWindowService,
+            IThreadingContext threadingContext,
+            IAsynchronousOperationListenerProvider listenerProvider)
+        {
+            _metadataAsSourceFileService = metadataAsSourceFileService;
+            _codeDefinitionWindowService = codeDefinitionWindowService;
+            _threadingContext = threadingContext;
+            _asyncListener = listenerProvider.GetListener(FeatureAttribute.CodeDefinitionWindow);
+        }
+
+        void ITextViewConnectionListener.SubjectBuffersConnected(ITextView textView, ConnectionReason reason, IReadOnlyCollection<ITextBuffer> subjectBuffers)
+        {
+            if (!_subscribedViews.Contains(textView) && !textView.Roles.Contains(PredefinedTextViewRoles.CodeDefinitionView))
+            {
+                _subscribedViews.Add(textView);
+                textView.Caret.PositionChanged += OnTextViewCaretPositionChanged;
+                QueueUpdateForCaretPosition(textView.Caret.Position);
+            }
+        }
+
+        void ITextViewConnectionListener.SubjectBuffersDisconnected(ITextView textView, ConnectionReason reason, IReadOnlyCollection<ITextBuffer> subjectBuffers)
+        {
+            if (reason == ConnectionReason.TextViewLifetime ||
+                !textView.BufferGraph.GetTextBuffers(b => b.ContentType.IsOfType(ContentTypeNames.RoslynContentType)).Any())
+            {
+                if (_subscribedViews.Contains(textView))
+                {
+                    _subscribedViews.Remove(textView);
+                    textView.Caret.PositionChanged -= OnTextViewCaretPositionChanged;
+                }
+            }
+        }
+
+        private void OnTextViewCaretPositionChanged(object? sender, CaretPositionChangedEventArgs e)
+        {
+            QueueUpdateForCaretPosition(e.NewPosition);
+        }
+
+        private void QueueUpdateForCaretPosition(CaretPosition caretPosition)
+        {
+            Contract.ThrowIfFalse(_threadingContext.JoinableTaskContext.IsOnMainThread);
+
+            // Cancel any pending update for this view
+            _currentUpdateCancellationToken?.Cancel();
+
+            // See if we moved somewhere else in a projection that we care about
+            var pointInRoslynSnapshot = caretPosition.Point.GetPoint(tb => tb.ContentType.IsOfType(ContentTypeNames.RoslynContentType), caretPosition.Affinity);
+            if (pointInRoslynSnapshot == null)
+            {
+                return;
+            }
+
+            _currentUpdateCancellationToken = new CancellationTokenSource();
+
+            var cancellationToken = _currentUpdateCancellationToken.Token;
+            var asyncToken = _asyncListener.BeginAsyncOperation(nameof(DefinitionContextTracker) + "." + nameof(QueueUpdateForCaretPosition));
+            UpdateForCaretPositionAsync(pointInRoslynSnapshot.Value, cancellationToken).CompletesAsyncOperation(asyncToken);
```

Comment:
so every caret move instantly kicks this off.  considering using a AsyncBatchingWorkQueue so you can throttle thigns if you get a lot of caret moves in a short time.

Reply:
@CyrusNajmabadi So AsyncBatchingWorkQueue isn't the right behavior really, it'll throttle but if I'm say holding down an arrow key for a few seconds, it'll still be firing off work in the middle, where I really just want to wait until the whole thing settles down. We've got ResettableDelay but I'm not sure that's much clearer.

There's other strangeness with AsyncBatchingWorkQueue: I really want it to just say "do the last one", but we don't have a way to kick out old work. The dedup mode doesn't guarantee ordering for purposes of "last", and I could disable dedup but then you're ending up with long lists to only look at the last one.

(no matter what the current code here is crappy since the Task.Delay that ensures there's a delay is buried way deep so moving that up)

Reply:
> and I could disable dedup but then you're ending up with long lists to only look at the last one.

True.  Though in practice is that a problem?  

--

If so, i do think that it would likely be trivial to have a knob/subclass of AsyncBatchingWorkQueue where you say you only want the last item added.    It not much of a 'queue' at that point of course. :)  maybe 'AsyncBatchingThrottle'?

Reply:
alternate design.  use the batchingqueue to throttle things. *then* when you execute asynchronously, quickly bounce back to UI thread to get the *current* caret pos instead of using what's in the queue.

Reply:
> Though in practice is that a problem?

My concern is if you're jumping around you're potentially rooting views and buffers which aren't exactly cheap. I think my takeaway here is AsyncBatchingWorkQueue is great, but this isn't the right place for it. I can made it work, but that's just as complicated than the pattern here.

Reply:
Going to leave this as is.

Reply:
> I really want it to just say "do the last one", but we don't have a way to kick out old work

Perhaps a [TaskDelayScheduler](https://github.com/dotnet/project-system/blob/0476235d70d1b65b8062492075db6fd86995612a/src/Microsoft.VisualStudio.ProjectSystem.Managed/Threading/Tasks/TaskDelayScheduler.cs). We already have the CancellationSeries that makes it work.

Reply:
@davidwengier: ah, there we go, that's what I want yes.

Reply:
https://drop.com/buy/stack-overflow-the-key-macropad